processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -147,31 +387,115 @@ public int indexOf(T value)
     return bufferIndexed.iterator();
   }
 
-  private final ByteBuffer theBuffer;
-  private final ObjectStrategy<T> strategy;
-  private final boolean allowReverseLookup;
-  private final int size;
+  public long getSerializedSize()
+  {
+    if (valueBuffers.size() != 1) {
+      throw new UnsupportedOperationException("Method not supported for version 2 GenericIndexed.");
+    }
+    return theBuffer.remaining() + 2 + 4 + 4;
+  }
 
-  private final int valuesOffset;
-  private final BufferIndexed bufferIndexed;
+  public void writeToChannel(WritableByteChannel channel) throws IOException
+  {
+    //version 2 will always have more than one buffer in valueBuffers.
+    if (valueBuffers.size() == 1) {
+      channel.write(ByteBuffer.wrap(new byte[]{VERSION_ONE, allowReverseLookup ? (byte) 0x1 : (byte) 0x0}));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(theBuffer.remaining() + 4)));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(size)));
+      channel.write(theBuffer.asReadOnlyBuffer());
+    } else {
+      throw new UnsupportedOperationException(
+          "GenericIndexed serialization for V2 is unsupported. Use GenericIndexedWriter instead.");
+    }
+  }
 
-  GenericIndexed(
-      ByteBuffer buffer,
-      ObjectStrategy<T> strategy,
-      boolean allowReverseLookup
-  )
+  /**
+   * Create a non-thread-safe Indexed, which may perform better than the underlying Indexed.
+   *
+   * @return a non-thread-safe Indexed
+   */
+  public GenericIndexed<T>.BufferIndexed singleThreaded()
   {
-    this.theBuffer = buffer;
-    this.strategy = strategy;
-    this.allowReverseLookup = allowReverseLookup;
+    final ByteBuffer copyHeaderBuffer = headerBuffer.asReadOnlyBuffer();
+
+    if (valueBuffers.size() == 1) {
+      final ByteBuffer copyBuffer = valueBuffers.get(0).asReadOnlyBuffer();
+      return new BufferIndexed()
+      {
+        @Override
+        public T get(final int index)
+        {
+          if (index < 0) {
+            throw new IAE("Index[%s] < 0", index);
+          }
+          if (index >= size) {
+            throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+          }
+
+          final int startOffset;
+          final int endOffset;
+
+          if (index == 0) {
```

Comment:
I kept both the blocks separately because version 1 does a check of `if (index == 0)` vs V2 check is ` if (index % numElementsPerValueFile == 0) `, we wanted to avoid `%` operation for version 1 indexed.